### PR TITLE
Rewrite Kernel#loop in Ruby

### PR DESCRIPTION
**Pull request by k0kubun**
because Ruby is faster than C ™️

## Interpreter

```
$ benchmark-driver -v --chruby 'before;after' benchmark/loop_generator.rb
before: ruby 3.2.0dev (2022-12-21T15:34:21Z master 398aaed2f0) [arm64-darwin22]
after: ruby 3.2.0dev (2022-12-21T18:57:36Z ruby-loop b860fd41db) [arm64-darwin22]
Calculating -------------------------------------
                         before       after
      loop_generator      0.888       0.943 i/s -       1.000 times in 1.125595s 1.060871s

Comparison:
                   loop_generator
               after:         0.9 i/s
              before:         0.9 i/s - 1.06x  slower
```

## YJIT

```
benchmark-driver -v --chruby 'before::before --yjit-call-threshold=1;after::after --yjit-call-threshold=1' benchmark/loop_generator.rb
before: ruby 3.2.0dev (2022-12-21T15:34:21Z master 398aaed2f0) +YJIT [arm64-darwin22]
after: ruby 3.2.0dev (2022-12-21T18:57:36Z ruby-loop b860fd41db) +YJIT [arm64-darwin22]
Calculating -------------------------------------
                         before       after
      loop_generator      0.962       1.146 i/s -       1.000 times in 1.039119s 0.872700s

Comparison:
                   loop_generator
               after:         1.1 i/s
              before:         1.0 i/s - 1.19x  slower
```

### CODE REVIEW

File: kernel.rb [hunk 1]
```diff
@@ -150,6 +150,44 @@ def yield_self
...
+  #      y << "two"
+  #      :ok
+  #    }
+  #
+  #    result = loop {
+  #      puts enum.next
+  #    } #=> :ok
+  def loop
+    unless Primitive.block_given_p
+      return Primitive.cexpr! 'SIZED_ENUMERATOR(self, 0, 0, rb_f_loop_size)'
```
Comment (paracycle) [inline kernel.rb:182]: What if we use `Enumerator.produce { nil }` here? Would that be slower than calling a primitive?

Comment (k0kubun) [reply kernel.rb]: We don't have a benchmark for this path yet, so I just chose the same way as `#then`, which seems least likely to slow down. We'd need a benchmark if we want to change the way things work here. Do you have an idea for benchmarking this in a realistic manner? Honestly I have never used `#loop` in a way that this path is used.

Comment (k0kubun) [reply kernel.rb]: Also, how is it supposed to be used? https://github.com/ruby/ruby/pull/6983/commits/5b56ab1b990bba7ac02f187feb2dbb9f2edb4198 didn't pass CI.

Comment (paracycle) [reply kernel.rb]: Weird, I can check the failure. Normally `Enumerator.produce { nil }` should return an endless stream of `nil` values, which is what a bare `loop` does, as well.

Comment (paracycle) [reply kernel.rb]: Ok, so apparently the enumerator returned by `loop` is supposed to yield nothing, not even `nil`, which is what the spec is testing for. On the other hand `Enumerator.produce { }` or `Enumerator.produce { nil }` always yields `nil` values.

What works though is:
```ruby
return enum_for(:loop) { Float::INFINITY }
```

Comment (k0kubun) [reply kernel.rb]: That worked. thank you!

Comment (eregon) [reply kernel.rb]: FYI TruffleRuby & Rubinius ave this figured out for a long time: https://github.com/oracle/truffleruby/blob/master/src/main/ruby/truffleruby/core/kernel.rb#L407-L418
Feel free to copy Ruby code from there.

Comment (k0kubun) [reply kernel.rb]: Nice, I didn't notice the rescue region can be limited to the latter portion.

**Pull request merged**